### PR TITLE
♻️ add osparc-gateway-server to repo (⚠️ DEVOPS)

### PR DESCRIPTION
**Pull request by sanderegg**
## What do these changes do?
**NOTE:** NO NEED TO REVIEW ALL FILES, they come from another repository.
Important files are:
- CI file
- docker-compose files
- Makefile

The repository containing the [osparc-dask-gateway](https://github.com/ITISFoundation/osparc-dask-gateway) needs to be tightly in sync with the osparc-simcore one and in particular with the dask-sidecar service.
Especially tests that rely on the dask-sidecar and its specific dask dependencies must be in sync. It is impossible to keep them in sync with all flavours (or at least very cumbersome).
Therefore it is moved into osparc-simcore repository.

- also brings partial implementation of Docker Compose V2

⚠️ DEVOPS:
- docker compose V2 brings some slight changes in how the stack file is created. officially it is the exact same but it is untrue. That means that docker-compose.yml contains some changes. It would be really good that @mrnicegyu11 @Surfict use their devops eyes to check if the corrections I did are compatible. Worst case we might to do some correction when this hits master.

## Related issue/s
- relates to https://github.com/ITISFoundation/osparc-issues/issues/885
- relates to https://github.com/ITISFoundation/osparc-issues/issues/675

## How to test

```bash
# start first simcore
made devenv
make .env
# ENSURE that WEBSERVER_DEV_FEATURES_ENABLED=1 and DIRECTOR_V2_DEV_FEATURES_ENABLED=1
make build up-prod
# start a local dask gateway server
cd services/osparc-gateway-server
make build up-prod
# the address, user and password of the gateway is printed in the terminal
```
Ensure the cluster is correctly setup in osparc:
image

Run a pipeline:
image

### CODE REVIEW

**Comment by mrnicegyu11** [comment]
Sorry at the moment I dont have the bandwidth to read all 4k line additions, can you please point me to the file/line where critical docker-compose changes are introduced and where devops shall check? thanks a lot!

**Comment by sanderegg** [comment]
> Sorry at the moment I dont have the bandwidth to read all 4k line additions, can you please point me to the file/line where critical docker-compose changes are introduced and where devops shall check? thanks a lot!

@mrnicegyu11 
https://github.com/ITISFoundation/osparc-simcore/pull/3960/files#diff-b97116d8659c501e784ac94482ba924e05f2fec49341c87abe5f8d4d4e3ffa5b
and BTW, FYI docker-compose V1 will not get any support any more from June.

**Comment by mrnicegyu11** [comment]
LGTM

File: .github/workflows/ci-testing-deploy.yml [hunk 1]
```diff
@@ -837,6 +842,60 @@ jobs:
...
+      - name: typecheck
+        run: |
+          source .venv/bin/activate && \
+          pushd services/osparc-gateway-server && \
+          make mypy
+      - name: test
+        run: |
+          source .venv/bin/activate && \
+          pushd services/osparc-gateway-server && \
+          make test-ci-unit
```
Comment (GitHK) [inline .github/workflows/ci-testing-deploy.yml:894]: We usually have .bash files with these instructions. Maybe create one here as well.

File: .github/workflows/ci-testing-deploy.yml [hunk 2]
```diff
@@ -1546,6 +1606,81 @@ jobs:
...
+      - name: integration-test
+        run: |
+          source .venv/bin/activate && \
+          pushd services/osparc-gateway-server && \
+          make test-ci-integration
+      - name: system-test
+        run: |
+          source .venv/bin/activate && \
+          pushd services/osparc-gateway-server && \
+          make test-system
```
Comment (GitHK) [inline .github/workflows/ci-testing-deploy.yml:1668]: same here

File: tests/environment-setup/test_used_docker_compose.py [hunk 3]
```diff
@@ -93,36 +92,62 @@ def ensure_env_file(env_devel_file: Path) -> Iterable[Path]:
...
+    docker_compose_config_script = (
+        osparc_simcore_scripts_dir / "docker" / "docker-compose-config.bash"
+    )
+    assert docker_compose_config_script.exists()
+    return docker_compose_config_script
+
+
 @pytest.mark.parametrize(
-    "compose_path", compose_paths, ids=lambda p: str(p.relative_to(repo_dir))
+    "compose_path", compose_paths, ids=str  # lambda p: str(p.relative_to(repo_dir))
```
Comment (GitHK) [inline tests/environment-setup/test_used_docker_compose.py:125]: It this comment stil relevant?

**Review by GitHK — APPROVED** [review]
Looks good to me. Just find some questions/suggestions below.

File: tests/environment-setup/test_used_docker_compose.py (hunk 3 above)
Comment (sanderegg) [reply tests/environment-setup/test_used_docker_compose.py:125]: good catch

File: .github/workflows/ci-testing-deploy.yml (hunk 1 above)
Comment (sanderegg) [reply .github/workflows/ci-testing-deploy.yml:894]: I actually would like to go away from these batch files.

File: .github/workflows/ci-testing-deploy.yml (hunk 2 above)
Comment (sanderegg) [reply .github/workflows/ci-testing-deploy.yml:1668]: same here

**Review by mrnicegyu11 — APPROVED** [review]
thanks a lot!

**Review by matusdrobuliak66 — APPROVED** [review]

File: services/osparc-gateway-server/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+
+WORKDIR /home/scu
+
+# bring installed package without build tools
+COPY --from=prod-only-deps --chown=scu:scu ${VIRTUAL_ENV} ${VIRTUAL_ENV}
+# copy docker entrypoint and boot scripts
+COPY --chown=scu:scu services/osparc-gateway-server/docker services/osparc-gateway-server/docker
+
+
+# TODO: Create healthcheck
```
Comment (pcrespov) [inline services/osparc-gateway-server/Dockerfile:140]: Commented code + TODO ?!

File: tests/environment-setup/test_used_docker_compose.py [hunk 5]
```diff
@@ -93,36 +92,62 @@ def ensure_env_file(env_devel_file: Path) -> Iterable[Path]:
...
     compose = yaml.safe_load(compose_path.read_text())
     print(
         str(compose_path.relative_to(repo_dir)), "-> version=", compose.get("version")
     )
 
     subprocess.run(
-        f"docker-compose --file {compose_path} --env-file {env_devel_file} --verbose config",
+        " ".join(
+            [
+                f"{docker_compose_config_bash}",
```
Comment (pcrespov) [inline tests/environment-setup/test_used_docker_compose.py:142]: thx!

File: services/docker-compose.yml [hunk 6]
```diff
@@ -598,6 +602,7 @@ networks:
       config:
         - subnet: "172.8.0.0/16"
   computational_services_subnet:
+    name: ${SWARM_STACK_NAME}_computational_services_subnet
```
Comment (pcrespov) [inline services/docker-compose.yml:605]: nice

File: services/docker-compose-deploy.yml [hunk 7]
```diff
@@ -30,6 +30,9 @@ services:
   migration:
     image: ${DOCKER_REGISTRY:-itisfoundation}/migration:${DOCKER_IMAGE_TAG:-latest}
 
+  osparc-gateway-server:
```
Comment (pcrespov) [inline services/docker-compose-deploy.yml:33]: so, is it now part of the stack? I thought that was not the use case?

**Review by pcrespov — APPROVED** [review]
nice!

File: services/docker-compose-deploy.yml (hunk 7 above)
Comment (sanderegg) [reply services/docker-compose-deploy.yml:33]: it is not part of the stack. but it is deployed to dockerhub

File: services/osparc-gateway-server/Dockerfile (hunk 4 above)
Comment (sanderegg) [reply services/osparc-gateway-server/Dockerfile:140]: yes this is bad. but was already there before.

**Comment by GitHK** [comment]
Question: @sanderegg does this mean that the docker volume names will change now?
I see this pattern `name: ${SWARM_STACK_NAME}_redis-data`. I guess you expect them not to change? Wanted to be sure about it.

@mrnicegyu11 @Surfict

**Comment by sanderegg** [comment]
@GitHK hopefully it should not. that is why I specifically already asked @mrnicegyu11 about it. we will see what happens in master.

**Pull request merged**